### PR TITLE
Deleting the connection owner? Think again (or transfer ownership.)

### PR DESCRIPTION
**Pull request by dereksmart**
#### Changes proposed in this Pull Request:

- Adds some new helper methods for getting info about the user connections.
- Modifies the permission callback for the `/connection/owner` endpoint.
- Adds a whole notice UI and logic during the delete user process with information and actions to take if the connection owner is being deleted.

**What this PR is intentionally not doing, and will be done in other PRs:**
- Add tracking
- Better copy/design

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* I wouldn't really call it a feature, more a preventative enhancement. 

#### Testing instructions:

There's a lot. 

**Test changing connection owner to secondary Admin**
1. Set up a new site. Connect Jetpack to whatever user - let's call it `User A`. 
2. Add a new admin user to the site - let'a call it `User B`.
3. In a separate (incognito?) window, log into the site as User B. 
4. Go to the Jetpack admin and connect User B to wordpress.com. 
5. Still logged in as User B, go to the users page. 
6. Click to delete User A. (Remember, User A should be the connection owner). You should be taken to a confirmation page to delete, where you should see this new notice: 
image
7. Since you are logged in as a secondary connected user, you should see a dropdown with your name in it. 
8. Click the button to switch the connection to you! 
9. The notice should disappear. 

**Verify it worked**
After you did the above, you should be able to verify it "worked" by doing the following things: 
- In CLI: `wp jetpack options get master_user` - should be User B's ID returned.
- In wordpress.com NA: Visit the blog's report card. You should see User B's wpcom connected user as the new owner. 
- You should also see the connection owner has been updated in the blog's audit trail in wpcom. 

**Test when there are no other users connected**
Since we're only able to transfer ownership to another connected user, we needed a case to handle if there were none! So, you should expect to see a lot of text and a button to connect if this is the case when you're trying to delete the connection owner.  For this one, please verify: 
- That connection link works, and is done in a new tab. 
- The support links work and points to the right places. 
image

**General things to check/verify**
- You should never see an option to transfer ownership to a non-admin
- You should never see an option to transfer ownership to another user that is not connected currently. 

#### Proposed changelog entry for your changes:

*General: Warn users that deleting the connection owner is going to break things.

### CODE REVIEW

File: packages/connection/src/Manager.php [hunk 1]
```diff
@@ -498,6 +498,31 @@ public function is_user_connected( $user_id = false ) {
...
 	}
 
+	/**
+	 * Returns an array of user_id's that have user tokens for communicating with wpcom.
+	 * Able to select by specific capability.
+	 *
+	 * @param string The capability of the user
+	 * @return array Array of WP_User objects if found.
+	 */
+	public function get_connected_users( $capability = 'any' ) {
```
Comment (dereksmart) [inline packages/connection/src/Manager.php]: Happy to break this out into it's own PR if it makes it easier.

File: packages/jitm/src/JITM.php [hunk 2]
```diff
@@ -80,6 +81,9 @@ function prepare_jitms( $screen ) {
 			add_action( 'admin_enqueue_scripts', array( $this, 'jitm_enqueue_files' ) );
 			add_action( 'admin_notices', array( $this, 'ajax_message' ) );
 			add_action( 'edit_form_top', array( $this, 'ajax_message' ) );
+
+			// Not really a JITM. Don't know where else to put this :)
```
Comment (dereksmart) [inline packages/jitm/src/JITM.php]: help :)

File: tests/php/_inc/lib/test_class.rest-api-endpoints.php [hunk 3]
```diff
@@ -933,7 +933,7 @@ public function test_change_owner() {
 
 		// Create a user and set it up as current.
 		$user = $this->create_and_get_user();
-		$user->add_cap( 'jetpack_connect_user' );
+		$user->add_cap( 'jetpack_disconnect' );
```
Comment (dereksmart) [inline tests/php/_inc/lib/test_class.rest-api-endpoints.php]: This was the wrong user cap to allow switching connection owner. Any user can connect_user, but not any user can own the connection.

File: packages/connection/src/Manager.php [hunk 4]
```diff
@@ -533,6 +558,22 @@ public function get_connected_user_data( $user_id = null ) {
...
+	 * Returns a user object of the connection owner.
+	 *
+	 * @return object|false False if no connection owner found.
+	 */
+	public function get_connection_owner() {
+		$user_token = $this->get_access_token( JETPACK_MASTER_USER );
+
+		$connection_owner = false;
+		if ( $user_token && is_object( $user_token ) && isset( $user_token->external_user_id ) ) {
+			$connection_owner = get_userdata( $user_token->external_user_id );
```
Comment (dereksmart) [inline packages/connection/src/Manager.php]: If the user is missing, this will break. I will update this to do the same as it's done in https://github.com/Automattic/jetpack/pull/13323 by returning ID only once it's merged.

File: tests/php/_inc/lib/test_class.rest-api-endpoints.php [hunk 5]
```diff
@@ -928,21 +928,18 @@ public function test_fetch_not_registered_widget_data() {
...
-		$user->add_cap( 'jetpack_connect_user' );
+		$user = $this->create_and_get_user( 'administrator' );
+		$user->add_cap( 'jetpack_disconnect' );
 		wp_set_current_user( $user->ID );
 
 		// Mock site already registered
 		Jetpack_Options::update_option( 'user_tokens', array( $user->ID => "honey.badger.$user->ID" ) );
 
-		// Attempt change, fail because not master user
-		$response = $this->create_and_get_request( 'connection/owner', array( 'owner' => 999 ), 'POST' );
```
Comment (dereksmart) [inline tests/php/_inc/lib/test_class.rest-api-endpoints.php]: The method now allows non-master users to use this endpoint.

File: packages/jitm/src/JITM.php [hunk 6]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+		_e( '<h2>Important notice about your Jetpack connection: </h2>', 'jetpack' );
+		printf(
+			__( "<p>Warning! You are about to delete the Jetpack <a href='%1\$1s' target='_blank'>connection owner</a> (%2\$2s) for this site, which may affect some of your features.</p>", 'jetpack' ),
+			'https://jetpack.com/support/primary-user/',
+			esc_html( $connection_owner->data->user_login )
+		);
+
+		if ( ! empty( $connected_users ) && count( $connected_users ) > 1 ) {
+			echo '<form id="jp-switch-connection-owner" action="" method="post">';
+			echo "<label for='owner'>" . __( 'You can choose to transfer connection ownership to one of these already-connected admins: ', 'jetpack' ) . '</label>';
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Empty trailing space in the translation string will likely get accidentally dropped by translators so if it's needed, better use CSS.

File: packages/jitm/src/JITM.php [hunk 7]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+				array(
+					'name'    => 'owner',
+					'include' => array_diff( $connected_user_ids, array( $connection_owner->ID ) ),
+					'show'    => 'display_name_with_login',
+				)
+			);
+
+			submit_button( __( 'Set new connection owner', 'jetpack' ), 'primary', 'jp-switch-connection-owner-submit' );
+
+			_e( '<p>As always, feel free to <a href="https://jetpack.com/contact-support" target="_blank">contact our support team</a> if you have any questions.</p>', 'jetpack' );
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Nit; this and other `target="_blank"` links would need `rel="noreferrer noopener"`

File: packages/jitm/src/JITM.php [hunk 8]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+		$deleting_connection_owner = in_array( $connection_owner->ID, (array) $user_ids_to_delete );
+		if ( ! $deleting_connection_owner ) {
+			return;
+		}
+
+		$connection_manager = new Manager();
+		$connected_users    = $connection_manager->get_connected_users( 'jetpack_connect' );
+
+		echo "<div class='notice notice-warning' id='jetpack-notice-switch-connection-owner'>";
+		_e( '<h2>Important notice about your Jetpack connection: </h2>', 'jetpack' );
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Would it be possible to leave out `<h2>` in the translation string?

File: packages/jitm/src/JITM.php [hunk 9]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+
+			echo "<div id='jp-switch-user-results'></div>";
+			echo '</form>';
+			?>
+			<script type="text/javascript">
+				jQuery( document ).ready( function( $ ) {
+					$( '#jp-switch-connection-owner' ).on( 'submit', function( e ) {
+						const formData = $( this ).serialize();
+						const submitBtn = document.getElementById( 'jp-switch-connection-owner-submit' );
+						const results = document.getElementById( 'jp-switch-user-results' );
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Since this is un-transpiled ES5 code, you should use `var` instead. (mind IE11)

Comment (simison) [reply packages/jitm/src/JITM.php]: ...tho maybe this bit could be in a `.js` file and enqueued instead so that we catch it better with linters and whatnot. :-)

File: packages/jitm/src/JITM.php (hunk 2 above)
Comment (simison) [reply packages/jitm/src/JITM.php]: As a dev, I'd probably search for something like this from `packages/connection` but not sure how this plays with that package needing to be more generic and not Jetpack-specific?

File: packages/jitm/src/JITM.php [hunk 10]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+	 * the connection owner.
+	 */
+	function delete_user_update_connection_owner_notice() {
+		global $current_screen;
+
+		if ( 'users' !== $current_screen->base ) {
+			return;
+		}
+
+		if ( 'delete' !== $_REQUEST['action'] ) {
```
Comment (simison) [inline packages/jitm/src/JITM.php]: You should check if `isset( $_REQUEST['action'] )` first to avoid _ Undefined index_ PHP notices.

File: packages/jitm/src/JITM.php [hunk 11]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+
+						$.ajax( {
+							type        : "POST",
+							url         : "<?php echo get_rest_url() . 'jetpack/v4/connection/owner'; ?>",
+							data        : formData,
+							headers     : {
+								'X-WP-Nonce': "<?php echo wp_create_nonce( 'wp_rest' ); ?>",
+							},
+							success: function() {
+								results.innerHTML = "Success! thank you.";
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Looks temporary but remember to pass via `__()` + `esc_js`. ;-)

File: packages/jitm/src/JITM.php [hunk 12]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+						e.preventDefault();
+						return false;
+					} );
+				} );
+			</script>
+			<?php
+		} else {
+			_e( '<p>Unfortunately, there are no other connected admins to transfer the connection to.</p>', 'jetpack' );
+			printf(
+				__( '<p>If you would like to be the new connection owner for this site, please connect to your WordPress.com account by clicking <a href="%s" target="_blank">this link</a>. Once you connect, you may refresh this page to see an option to change the connection owner.</p>', 'jetpack' ),
```
Comment (simison) [inline packages/jitm/src/JITM.php]: What do you think if you'd move the link from "this link" to "connect..."?

> If you would like to be the new connection owner for this site, please connect to your WordPress.com account by clicking [this link].
→
> If you would like to be the new connection owner for this site, please [connect to your WordPress.com account].

File: packages/jitm/src/JITM.php [hunk 13]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+				\Jetpack::init()->build_connect_url( false, false, 'connection_owner_notice' )
+			);
+			_e( '<p>If the site does not have a connection owner, it will be disconnected from Jetpack servers.</p>', 'jetpack' );
```
Comment (simison) [inline packages/jitm/src/JITM.php]: "Jetpack servers" sounds very technical, maybe there's a more understandable way to describe the significance of disconnection?

File: packages/jitm/src/JITM.php (hunk 7 above)
Comment (simison) [reply packages/jitm/src/JITM.php]: No idea if this is possible easily but would be nice to have "external" icon after `_blank` links.

File: packages/jitm/src/JITM.php [hunk 14]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+		// Get connection owner or bail
+		$connection_manager = new Manager();
+		$connection_owner   = $connection_manager->get_connection_owner();
+		if ( ! $connection_owner ) {
+			return;
+		}
+
+		// Bail if we're not trying to delete connection owner.
+		$user_ids_to_delete        = $_REQUEST['users'];
+		$deleting_connection_owner = in_array( $connection_owner->ID, (array) $user_ids_to_delete );
```
Comment (simison) [inline packages/jitm/src/JITM.php]: Instead of adding `get_connection_owner()` method, would looping `$user_ids_to_delete` using existing `$connection_manager->is_connection_owner()` method be sufficient enough?

The method sounds fairly useful in a connection package tho. :-)

**Review by simison — COMMENTED** [review]
I really like this idea so I went ahead and left a few minor comments as a review, hope this helps!

Not necessarily in this PR but might make sense to hide the notice when user is deleting themself and WP is stopping them:

Screenshot 2019-08-26 at 09 54 43


Otherwise the notice might confuse.

File: packages/jitm/src/JITM.php [hunk 15]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+
+		// Get connection owner or bail
+		$connection_manager = new Manager();
+		$connection_owner   = $connection_manager->get_connection_owner();
+		if ( ! $connection_owner ) {
+			return;
+		}
+
+		// Bail if we're not trying to delete connection owner.
+		$user_ids_to_delete        = $_REQUEST['users'];
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think it should be `user` here:

```suggestion
		$user_ids_to_delete        = $_REQUEST['user'];
```

File: packages/jitm/src/JITM.php [hunk 16]
```diff
@@ -149,6 +153,116 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+			return;
+		}
+
+		$connection_manager = new Manager();
+		$connected_users    = $connection_manager->get_connected_users( 'jetpack_connect' );
+
+		echo "<div class='notice notice-warning' id='jetpack-notice-switch-connection-owner'>";
+		_e( '<h2>Important notice about your Jetpack connection: </h2>', 'jetpack' );
+		printf(
+			__( "<p>Warning! You are about to delete the Jetpack <a href='%1\$1s' target='_blank'>connection owner</a> (%2\$2s) for this site, which may affect some of your features.</p>", 'jetpack' ),
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: "your site's features", maybe?

File: packages/jitm/src/JITM.php (hunk 6 above)
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I wonder if we should have a different wording when there is only one other admin, the person that's going through the flow?

File: packages/jitm/src/JITM.php (hunk 12 above)
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I wonder if we should be more imperative in this case? It's not just "if you would like", they really need to do this or they'll lose features. The connect button (I think it should be a button) should become the main CTA on the page imo.

**Review by jeherve — CHANGES_REQUESTED** [review]
I like the idea; I think it'd be a great addition to the Users screen. 👍

File: packages/jitm/src/JITM.php (hunk 2 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: We decided at one point to let UI decisions be made by the consumer, and not include UI into the connection package. We can decide the other way later, but it's not important enough to block this PR over IMO.

File: packages/jitm/src/JITM.php (hunk 9 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: `const` looks ok for IE11 except in loops? https://caniuse.com/#feat=const

File: packages/jitm/src/JITM.php (hunk 13 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: "Some Jetpack features may stop working."

File: packages/jitm/src/JITM.php (hunk 14 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: What if the site has 10,000 connected users? (I know unlikely, but this is more performant in scale.)

> The method sounds fairly useful in a connection package tho. :-)

I thought so too and was surprised it didn't already exist.

File: packages/jitm/src/JITM.php (hunk 15 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Turns out it's both :) 

It's `user` if you are only deleting a single user with the `delete` button straight away, but if you use the checkbox/bulk action, then it's `users` always. I will add a check for `user` as well! good catch.

File: packages/jitm/src/JITM.php (hunk 6 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Honestly, it's too wordy for me. I think the better way to approach it is to keep it simple with something like "Your site may function properly", and update/link to a support page.

Comment (dereksmart) [reply packages/jitm/src/JITM.php]: fixed!

File: packages/jitm/src/JITM.php (hunk 7 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Done for all, thanks!

File: packages/jitm/src/JITM.php (hunk 9 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: I updated it anyway

File: packages/jitm/src/JITM.php (hunk 11 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Yes it's temporary

File: packages/jitm/src/JITM.php (hunk 15 above)
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Added a check for both!

**Comment by dereksmart** [comment]
I would love if choice of words/design does not block this PR. What do you think about simplifying the wording and instead just linking to a support page such as this where we can add more information? https://jetpack.com/support/primary-user/

File: packages/jitm/src/JITM.php (hunk 2 above)
Comment (simison) [reply packages/jitm/src/JITM.php]: Makes total sense! 👍

**Comment by dereksmart** [comment]
Made some adjustments to the copy and layout. I think it's clearer now, but we'll need to update that support page during beta. 

image

image

Ready for review again.

File: packages/jitm/src/JITM.php (hunk 7 above)
Comment (simison) [reply packages/jitm/src/JITM.php]: Yup 👍  Totally not a blocker, I did mean to mention that. :-) Lots of places in Jetpack seem to miss it already, too.

File: packages/jitm/src/JITM.php (hunk 9 above)
Comment (simison) [reply packages/jitm/src/JITM.php]: IE11 keeps surprising sometimes! 😍

File: packages/jitm/src/JITM.php [hunk 17]
```diff
@@ -149,6 +153,152 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+			return;
+		}
+
+		// Bail if they're trying to delete themselves to avoid confusion.
+		if ( $connection_owner->ID == get_current_user_id() ) {
+			return;
+		}
+
+		$connection_manager = new Manager();
+		$connected_users    = $connection_manager->get_connected_users( 'jetpack_connect' );
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: Should it be `$connnected_admins` here, since we are only looking for users that can connect Jetpack to WordPress.com for the site? I was a bit confused at first.

File: packages/jitm/src/JITM.php [hunk 18]
```diff
@@ -149,6 +153,152 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+
+		echo "<div class='notice notice-warning' id='jetpack-notice-switch-connection-owner'>";
+		echo '<h2>' . esc_html__( 'Important notice about your Jetpack connection:', 'jetpack' ) . '</h2>';
+		echo '<p>' . sprintf( esc_html__( "Warning! You are about to delete the Jetpack connection owner (%s) for this site, which may cause some of your Jetpack features to stop working.", 'jetpack' ),
+			esc_html( $connection_owner->data->user_login )
+		) . '</p>';
+
+		if ( ! empty( $connected_users ) && count( $connected_users ) > 1 ) {
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think using `$connnected_admins` instead of `$connected_users` would be especially useful here. At first I didn't get why we needed to have **more than** 1 connected user. But in fact we need at least 2 connected admins: the current master user and the future one.

File: packages/connection/src/Manager.php (hunk 4 above)
Comment (jeherve) [reply packages/connection/src/Manager.php]: #13323 has now been merged.

File: packages/jitm/src/JITM.php [hunk 19]
```diff
@@ -149,6 +153,152 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+			echo '<form id="jp-switch-connection-owner" action="" method="post">';
+			echo "<label for='owner'>" . esc_html__( 'You can choose to transfer connection ownership to one of these already-connected admins:', 'jetpack' ) . ' </label>';
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: Not a blocker, but I think in a future PR we should also mention something about the plan, when the site has one. We want to let folks know that the plan will remain attached the the previous owner.

File: packages/jitm/src/JITM.php [hunk 20]
```diff
@@ -149,6 +153,152 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+						e.preventDefault();
+						return false;
+					} );
+				} );
+			</script>
+			<?php
+		} else {
+			echo '<p>' . esc_html__( 'Every Jetpack site needs at least one connected admin for the features to work properly. Please connect to your WordPress.com account via the button below. Once you connect, you may refresh this page to see an option to change the connection owner.', 'jetpack' ) . "</p>";
+			$connect_url = \Jetpack::init()->build_connect_url( false, false, 'connection_owner_notice' );
+			echo "<a href='" . esc_url( $connect_url ) . "' target='_blank' rel='noopener noreferrer' class='button-primary'>" . esc_html__( 'Connect to Wordpress.com' ) . "</a>";
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think you're missing a texdomain here.

```suggestion
			echo "<a href='" . esc_url( $connect_url ) . "' target='_blank' rel='noopener noreferrer' class='button-primary'>" . esc_html__( 'Connect to WordPress.com', 'jetpack' ) . "</a>";
```

**Review by jeherve — DISMISSED** [review]

File: packages/jitm/src/JITM.php [hunk 21]
```diff
@@ -149,6 +153,153 @@ static function jitm_jetpack_woo_services_activate( $CTA ) {
...
+				function( $connected_admin ) {
+						return $connected_admin->ID;
+				},
+				$connected_admins
+			);
+
+			wp_dropdown_users(
+				array(
+					'name'    => 'owner',
+					'include' => array_diff( $connected_admin_ids, array( $connection_owner->ID ) ),
```
Comment (jeherve) [inline packages/jitm/src/JITM.php]:
```suggestion
					'include' => array_diff( $connected_admin_ids, array( $connection_owner_id ) ),
```

File: packages/jitm/src/JITM.php (hunk 19 above)
Comment (jeherve) [reply packages/jitm/src/JITM.php]: Only checking for `isset` may not be enough here since `get_userdata( $connection_owner_id )` will always return at least `false`. How about this instead:

```suggestion
			is_a( $connection_owner_userdata, 'WP_User' ) ? esc_html( $connection_owner_userdata->data->user_login ) : ''
```

**Review by jeherve — CHANGES_REQUESTED** [review]

**Review by jeherve — DISMISSED** [review]
This should be good to go like this! 👍

**Review by jeherve — APPROVED** [review]
